content/post/2019-12-10-pkg-urls.Rmarkdown

```diff
@@ -0,0 +1,159 @@
+
+* in the GitHub repo URL field, or the equivalent for your development platform?
+
+Have you used any of your package's URLs
+
+* [In your public message about your package, e.g. as an answer to someone's question](https://community.rstudio.com/t/pkgdown-site-seo/26706/6)?
+
+* In the slides of [your talk about the package](https://www.tidyverse.org/blog/2018/07/carpe-talk/)?
+
+Don't miss any opportunity to point users and contributors in the right direction!
+
+## What URLs in DESCRIPTION files of CRAN packages?
+
+In the following, we shall parse the URL field of the CRAN packages database.
+
+```{r urls1}
+db <- tools::CRAN_package_db()
+
+db <- tibble::as_tibble(db[, c("Package", "URL")])
+db <- dplyr::distinct(db)
+```
+
+There are `r nrow(db)` packages on CRAN at the time of writing, among which `r sum(is.na(db$URL))` with something written in the URL field. We can parse this data.
+
+```{r urls2}
+db <- db[!is.na(db$URL),]
+
+library("magrittr")
+
+# function from https://github.com/r-hub/pkgsearch/blob/26c4cc24b9296135b6238adc7631bc5250509486/R/addin.R#L490-L496
+
+url_regex <- function() "(https?://[^\\s,;>]+)"
+
+find_urls <- function(txt) {
+  mch <- gregexpr(url_regex(), txt, perl = TRUE)
+  res <- regmatches(txt, mch)[[1]]
+
+  if(length(res) == 0) {
+    return(list(NULL))
+  } else {
+    list(unique(res))
+  }
+}
+
+db %>%
+  dplyr::group_by(Package)  %>%
+  dplyr::mutate(actual_url = find_urls(URL))%>%
+  dplyr::ungroup() %>%
+  tidyr::unnest(actual_url) %>%
+  dplyr::group_by(Package, actual_url) %>%
+  dplyr::mutate(url_parts = list(urltools::url_parse(actual_url))) %>%
+  dplyr::ungroup() %>%
+  tidyr::unnest(url_parts) %>%
+  dplyr::mutate(scheme = trimws(scheme)) -> parsed_db
+```
+
+There are `r length(unique(parsed_db$Package))` with at least one valid URL.
+
+What are the packages with most links?
+
+```{r urlsno}
+mostlinks <- dplyr::count(parsed_db, Package, sort = TRUE)
+mostlinks
+```
+
+The package with the most links in `URL` is `r gluedown::md_link(mostlinks$Package[1], glue::glue(" https://CRAN.R-project.org/package={mostlinks$Package[1]}"))`.
+
+What is the most popular scheme, [http or https](https://howhttps.works/)? 
+
+```{r scheme}
+dplyr::count(parsed_db, scheme, sort = TRUE)
+```
+
+There is a bit less that one third of http links.
+
+Can we identify popular domains?
+
+```{r domain}
+dplyr::count(parsed_db, domain, sort = TRUE)
+
+```
+
+GitHub seems to be the most popular development platform, as least from this sample of CRAN packages that indicate an URL. It is also possible that some developers set up their own GitLab server with a own domain. 
+Many packages link to `www.r-project.org` which is not very informative, or to their own CRAN page which can be informative. 
+
+> By the way if you visit CRAN pages to get package metadata such as dependencies or reverse dependencies, you could also use [our `pkgsearch` package](/2019/11/29/pkgsearch/).
```

Comment:
Hmmm, this dependency part seems to be a bit out of place to me. Is it actually needed for the rest of post?

Reply:
right, no. btw I opened an issue in pkgsearch about that.